base_layer/wallet/src/text_message_service/model.rs

```diff
@@ -0,0 +1,661 @@
+/// send a duplicate message to the same destination that the ID hashes will be unique
+pub fn generate_id<D: Digest>(
+    source_pub_key: &CommsPublicKey,
+    dest_pub_key: &CommsPublicKey,
+    message: &String,
+    timestamp: &NaiveDateTime,
+    index: usize,
+) -> Vec<u8>
+{
+    D::new()
+        .chain(source_pub_key.as_bytes())
+        .chain(dest_pub_key.as_bytes())
+        .chain(message.as_bytes())
+        .chain(timestamp.to_string())
+        .chain(index.to_le_bytes())
+        .result()
+        .to_vec()
+}
+
+/// Represents a single Text Message to be sent that includes an acknowledged field
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct SentTextMessage {
+    pub id: Vec<u8>,
+    pub source_pub_key: CommsPublicKey,
+    pub dest_pub_key: CommsPublicKey,
+    pub message: String,
+    pub timestamp: NaiveDateTime,
+    pub acknowledged: bool,
+}
+
+/// The Native Sql version of the SentTextMessage model
+#[derive(Insertable, Queryable)]
+#[table_name = "sent_messages"]
+pub struct SentTextMessageSql {
+    pub id: Vec<u8>,
+    pub source_pub_key: String,
+    pub dest_pub_key: String,
+    pub message: String,
+    pub timestamp: NaiveDateTime,
+    pub acknowledged: i32,
+}
+
+impl SentTextMessage {
+    pub fn new(source_pub_key: CommsPublicKey, dest_pub_key: CommsPublicKey, message: String) -> SentTextMessage {
+        let timestamp = Utc::now().naive_utc();
+        // TODO implement automatic calculation of the index of messages sent to this recipient to ensure that rapid
+        // duplicate messages have different IDs
+        let id = generate_id::<HashDigest>(&source_pub_key, &dest_pub_key, &message, &timestamp, 0);
+        SentTextMessage {
+            id,
+            source_pub_key,
+            dest_pub_key,
+            message,
+            timestamp,
+            acknowledged: false,
+        }
+    }
+
+    pub fn commit(&self, conn: &SqliteConnection) -> Result<(), TextMessageError> {
+        diesel::insert_into(sent_messages::table)
+            .values(SentTextMessageSql::from(self.clone()))
+            .execute(conn)?;
+        Ok(())
+    }
+
+    pub fn find(id: &Vec<u8>, conn: &SqliteConnection) -> Result<SentTextMessage, TextMessageError> {
+        SentTextMessage::try_from(
+            sent_messages::table
+                .filter(sent_messages::id.eq(id))
+                .first::<SentTextMessageSql>(conn)?,
+        )
+    }
+
+    pub fn find_by_dest_pub_key(
+        dest_pub_key: &CommsPublicKey,
+        conn: &SqliteConnection,
+    ) -> Result<Vec<SentTextMessage>, TextMessageError>
+    {
+        let mut result = sent_messages::table
+            .filter(sent_messages::dest_pub_key.eq(dest_pub_key.to_hex()))
+            .order_by(sent_messages::timestamp)
+            .load::<SentTextMessageSql>(conn)?;
+        let mut deserialized: Vec<Result<SentTextMessage, TextMessageError>> =
```

Comment:
In this case it may be better to use a for loop to prevent extra iterations i.e add the converted struct to a vec and return an error as soon as one fails. Same for `index()` below and a few other places

Reply:
Agreed, the lure of fancy functional stuff is strong. I will do this in the next PR.